### PR TITLE
feat: Add the ability to send HTTP requests

### LINKED ISSUE #12: (Dogfooding) Integrate CI server with GH repo so that it runs automatically
This is just about having the CI server run without doing any checks at all, and always passing.

[This guide from GitHub shows how to set this up](https://docs.github.com/en/developers/apps/guides/creating-ci-tests-with-the-checks-api)

### LINKED ISSUE #28: Send OK message to GitHub for any push
To close the pipeline, we should make let the CI answer with some sort of "OK" message for every push. This way, we will have an easier time dogfooding the solution in the future.

Related issue #14 for the finished implementation.

### PR DESCRIPTION
**Pull request by danielwis**
## Describe your changes
This adds `send` functionality to the `PipelineUpdateRequest` class, used to send the actual HTTP POST request and update the commit status. Fixes #12 (also fixes #28; these are possible duplicates).

## Checklist before requesting review

-   [x] Feature/fix PRs should add one atomic (as small as possible) feature or
        fix
-   [x] If your PR adds a feature or fix it also needs to add or modify at least
        one test
-   [x] All code in your PR needs to have been formatted
-   [x] The merge commit title should follow our commit prefix convention of
        either "feat", "fix", "chore", "doc", or "refactor"
-   [x] The merge commit body should reference at least one issue
-   [x] The code compiles and all the tests pass

### CODE REVIEW

File: app/src/main/java/ciserver/ContinuousIntegrationServer.java [hunk 1]
```diff
@@ -18,6 +19,7 @@
 public class ContinuousIntegrationServer extends AbstractHandler {
 
     Gson gson = new Gson();
```
Comment (didrikmunther) [inline app/src/main/java/ciserver/ContinuousIntegrationServer.java]: Can this also be final? Considering it should never change. Not terribly important

File: app/src/main/java/ciserver/ContinuousIntegrationServer.java [hunk 2]
```diff
@@ -18,6 +19,7 @@
...
+    final String GH_ACCESS_TOKEN = System.getenv("GH_ACCESS_TOKEN");
```
Comment (psoder) [inline app/src/main/java/ciserver/ContinuousIntegrationServer.java]:
```suggestion
    private Gson gson = new Gson();
    private final String GH_ACCESS_TOKEN = System.getenv("GH_ACCESS_TOKEN");
```

File: app/src/main/java/ciserver/PipelineUpdateRequest.java [hunk 3]
```diff
@@ -28,8 +36,9 @@ class RequestPath {
...
 }
 
-// Body parameters. If this class is needed for another
-// type of post request later on, we could add more.
+/**
+ * The RequestBody contains the body parameters of the PipelineUpdateRequest.
+ **/
 class RequestBody {
     String state;
     String target_url;
```
Comment (didrikmunther) [inline app/src/main/java/ciserver/PipelineUpdateRequest.java]: You can use `@SerializedName("target_url") String targetUrl` in order to comply with Java naming.

File: app/src/main/java/ciserver/PipelineUpdateRequest.java [hunk 4]
```diff
@@ -44,16 +53,70 @@ class RequestBody {
...
+ *
+ * Currently, the GitHub API version used is hard-coded to 2022-11-28.
+ **/
 public class PipelineUpdateRequest {
+    static final String API_VERSION = "2022-11-28";
+
     RequestPath path;
     RequestBody body;
     Optional<PipelineStage> failedOn;
+    String auth_tkn;
```
Comment (didrikmunther) [inline app/src/main/java/ciserver/PipelineUpdateRequest.java]: `authTkn`

File: app/src/main/java/ciserver/PipelineUpdateRequest.java [hunk 5]
```diff
@@ -44,16 +53,70 @@ class RequestBody {
...
+    /**
+     * Sends an HTTP POST request to the repository specified in the
+     * PipelineUpdateRequest path,
+     * containing information about the success status of a certain build.
+     *
+     * @return the HTTP response formatted as a String.
+     **/
+    public HttpResponse<String> send() throws InterruptedException, IOException {
+        Gson gson = new Gson();
+        String request_body = gson.toJson(this.body);
```
Comment (didrikmunther) [inline app/src/main/java/ciserver/PipelineUpdateRequest.java]: `requestBody`

File: app/src/main/java/ciserver/PipelineUpdateRequest.java [hunk 6]
```diff
@@ -44,16 +53,70 @@ class RequestBody {
...
+     * @param url      A link to view build logs and other output from the server.
+     * @param desc     A short description of the state.
+     * @param ctx      A (case insensitive) label used to differentiate this state
+     *                 from a state set by another system.
+     * @param failedOn The stage where the Pipeline failed, or null if it succeeded.
+     **/
+    PipelineUpdateRequest(String owner, String repo, String sha, String token, CommitStatus state, String url,
+            String desc, String ctx, PipelineStage failedOn) {
         this.path = new RequestPath(owner, repo, sha);
         String state_str = state.toString().toLowerCase();
```
Comment (didrikmunther) [inline app/src/main/java/ciserver/PipelineUpdateRequest.java]: `stateStr`

File: app/src/test/java/ciserver/PipelineUpdateRequestTest.java [hunk 7]
```diff
@@ -3,35 +3,55 @@
...
+        PipelineUpdateRequest pr = new PipelineUpdateRequest("soffan-group-25", "ci-server", "123abc", "token",
+                CommitStatus.SUCCESS, "www.learn-more-about.this-build.com", "Test passed!", "ci", null);
+
+        try {
+        HttpResponse<String> resp = pr.send();
+        // API response code for "bad credentials"
+        assert(resp.statusCode() == 401);
+        // "message": "Bad credentials"
+        assert(resp.body().contains("Bad credentials"));
+        } catch (Exception e) {}
```
Comment (didrikmunther) [inline app/src/test/java/ciserver/PipelineUpdateRequestTest.java]: Should this test fail if there was an exception?

**Review by didrikmunther — COMMENTED** [review]
Looks good, some variable names might need changing.

File: app/src/main/java/ciserver/PipelineUpdateRequest.java [hunk 8]
```diff
@@ -44,16 +53,70 @@ class RequestBody {
...
+     * @param token    An API authentication token used by the CI server to access
+     *                 the repository and set the commit status.
+     * @param state    The state to set the commit status to.
+     * @param url      A link to view build logs and other output from the server.
+     * @param desc     A short description of the state.
+     * @param ctx      A (case insensitive) label used to differentiate this state
+     *                 from a state set by another system.
+     * @param failedOn The stage where the Pipeline failed, or null if it succeeded.
+     **/
+    PipelineUpdateRequest(String owner, String repo, String sha, String token, CommitStatus state, String url,
```
Comment (psoder) [inline app/src/main/java/ciserver/PipelineUpdateRequest.java]: I think it would be better to pass the parameters as a DTO.

**Review by psoder — CHANGES_REQUESTED** [review]

File: app/src/main/java/ciserver/PipelineUpdateRequest.java (hunk 8 above)
Comment (psoder) [reply app/src/main/java/ciserver/PipelineUpdateRequest.java]: https://www.baeldung.com/java-dto-pattern

**Review by psoder — APPROVED** [review]

**Pull request merged**